allenact/base_abstractions/distributions.py

```diff
@@ -71,6 +80,322 @@ def probs_tensor(self):
+        )
+
+    def sample(self, sample_shape=torch.Size(), **kwconds):
+        self.get_subpolicy(**kwconds)
+        return self.subpolicy.sample(sample_shape)
+
+    def mode(self, **kwconds):
+        self.get_subpolicy(**kwconds)
+        return self.subpolicy.mode()
+
+
+class DirectedGraphicalModel(Distr):
+    def __init__(self, cond_distrs: Sequence[CondDistr]):
+        self.partials: List[CondDistr]
+        self.deps: Dict[Tuple[str], List[Tuple[str]]]
+        self.partials, self.deps = DirectedGraphicalModel.make_ordered_partials(
+            cond_distrs
+        )
+        assert len(self.partials) == len(
+            cond_distrs
+        ), f"Inconsistent number of used ({len(self.partials)}) and given ({len(cond_distrs)}) partial distributions"
+        self.action_spaces = [partial.action_space for partial in self.partials]
+
+    @staticmethod
+    def collapse_deps(
+        varnames: Sequence[str], who_produces: Dict[str, Tuple[str]]
+    ) -> List[Tuple[str, ...]]:
+        """Generates a list of produced partial distributions required by the list of varnames.
+        """
+        current_partials = set()
+        for d in varnames:
+            current_partials.add(who_produces[d])
+        return list(current_partials)
+
+    @staticmethod
+    def make_ordered_partials(
+        cond_distrs: Sequence[CondDistr],
+    ) -> Tuple[List[CondDistr], Dict[Tuple[str], List[Tuple[str]]]]:
+        """Generates a computation order given a list of `CondDistr`.
+           It also provides a dict with produced partial distributions required by the list of produced vars.
+        """
+        producer: Dict[Tuple[str, ...], CondDistr] = {
+            tuple(d.produces): d for d in cond_distrs
+        }
+        who_produces: Dict[str, Tuple[str]] = {}
+        deps: Dict[Tuple[str], List[Union[str, Tuple[str]]]] = {
+            tuple(d.produces): list(d.given) for d in cond_distrs if d.given is not None
+        }
+        for partial in cond_distrs:
+            assert (
+                partial.produces is not None
+            ), "produces must be provided for cond_distrs"
+            for varname in partial.produces:
+                assert (
+                    varname not in who_produces
+                ), "'{}' is a duplicated produced variable".format(varname)
+                who_produces[varname] = tuple(partial.produces)
+
+        g = nx.DiGraph()
+        for k in deps:
+            deps[k] = DirectedGraphicalModel.collapse_deps(
+                deps[k], who_produces
+            )  # List[str] -> List[Tuple[str]]
+            for j in deps[k]:
+                g.add_edge(j, k)
+
+        assert nx.is_directed_acyclic_graph(
+            g
+        ), "partial distributions do not form a direct acyclic graph"
+
+        # ensure dependencies are precomputed
+        return [producer[n] for n in nx.dfs_postorder_nodes(g)][::-1], deps
+
+    def entropy(self):
```

Comment:
It's worth noting that this isn't actually the entropy of the joint distribution of the graphical model (intuitively this makes sense we're not taking into account the entropy of the conditional distributions had they been conditioned on something else). I don't think this matters in practice (in fact this version of entropy has some nice properties associated with it).

I'm not it's computationally feasible to compute the true entropy in general. One way around this might instead be to just approximate it by sampling using that entropy = -E_p[log p].